internal/instrumentation/instrumentation.go

```diff
@@ -0,0 +1,294 @@
+
+	if err := srv.ListenAndServe(); err != nil && !errors.Is(err, net.ErrClosed) {
+		return err
+	}
+
+	return nil
+}
+
+func (m *MetricsServer) auditCpuWorker(ctx context.Context) {
+	var lastIdle uint64 = 0
+	var lastTotal uint64 = 0
+
+	ticker := time.NewTicker(5 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			m.log.Debug("Stopping CPU audit")
+			return
+		case <-ticker.C:
+			stats, err := cpu.Get()
+			if err != nil {
+				m.log.Errorf("Could not audit cpu usage: %v", err)
+			}
+
+			// stats from /proc/stat increase monotonically, so we must
+			// compute the delta from our last audit
+			m.metrics.CpuUtilization.Set(
+				float64(stats.Idle-lastIdle) / float64(stats.Total-lastTotal),
+			)
+			lastIdle = stats.Idle
+			lastTotal = stats.Total
+		}
+	}
+}
+
+func (m *MetricsServer) auditMemoryWorker(ctx context.Context) {
+	ticker := time.NewTicker(time.Second * 5)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			m.log.Debug("Stopping memory audit")
+			return
+		case <-ticker.C:
+			stats, err := memory.Get()
+			if err != nil {
+				m.log.Errorf("could not audit memory usage: %v", err)
+			}
+
+			m.metrics.MemoryUtilization.Set(
+				float64(stats.Used) / float64(stats.Total),
+			)
+		}
+	}
+}
+
+func (m *MetricsServer) auditDiskWorker(ctx context.Context) {
+	ticker := time.NewTicker(time.Second * 5)
+	defer ticker.Stop()
+
+	var stat unix.Statfs_t
+
+	for {
+		select {
+		case <-ctx.Done():
+			m.log.Debug("Stopping disk audit")
+			return
+		case <-ticker.C:
+			err := unix.Statfs("/", &stat)
+			if err != nil {
+				fmt.Println("could not audit disk usage: ", err)
+			}
+
+			m.metrics.DiskUtilization.Set(
+				float64(stat.Bfree) / float64(stat.Blocks),
+			)
+		}
+	}
+}
+
+func (m *ApiMetrics) RegisterWith(reg *prometheus.Registry) {
+	reg.MustRegister(m.SuccessLatency)
+	reg.MustRegister(m.ErrorLatency)
+	reg.MustRegister(m.ApiTraffic)
+	reg.MustRegister(m.AgentTraffic)
+	reg.MustRegister(m.SloViolations)
+	reg.MustRegister(m.ServerErrors)
+	reg.MustRegister(m.CpuUtilization)
+	reg.MustRegister(m.MemoryUtilization)
+	reg.MustRegister(m.DiskUtilization)
+
+}
+
+func (m *ApiMetrics) AgentServerMiddleware(next http.Handler) http.Handler {
+	return m.ServerMiddleware(next, false)
+}
+
+func (m *ApiMetrics) ApiServerMiddleware(next http.Handler) http.Handler {
+	return m.ServerMiddleware(next, true)
+}
+
+func (m *ApiMetrics) ServerMiddleware(next http.Handler, agentServer bool) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		start := time.Now()
+
+		if agentServer {
+			m.AgentTraffic.Inc()
+		} else {
+			m.ApiTraffic.Inc()
+		}
+
+		lw := NewLoggingResponseWriter(w)
+		next.ServeHTTP(lw, r)
+		statusClass := lw.statusCode - lw.statusCode%100
+
+		if statusClass == 400 {
+			m.ClientErrors.Inc()
+		}
+
+		if statusClass == 500 {
+			m.ServerErrors.Inc()
+		}
+
+		stop := time.Now()
+		thisLatency := stop.Sub(start).Seconds()
```

Comment:
Are we sure seconds is granular enough?

```suggestion
thisLatency := time.Since(start).Seconds()
```